infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -580,82 +580,175 @@ private LogData getLogData(LogEntry entry) {
+     * partially written.
+     *
+     * @param ch
+     * @return
+     * @throws IOException
+     */
+    private LogHeader parseHeader(FileChannel ch) throws IOException {
+        Metadata metadata = parseMetadata(ch);
+        if (metadata == null) {
+            // Partial write on the metadata for the header
+            // Rewind the channel position to the begining of the file
+            ch.position(0);
+            return null;
+        }
 
-        fc.position(fc.position() + headerMetadata.getLength());
-        long channelOffset = fc.position();
-        ByteBuffer o = ByteBuffer.allocate((int) logFileSize - (int) fc.position());
-        fc.read(o);
-        fc.close();
-        o.flip();
+        ByteBuffer buffer = getPayloadForMetadata(ch, metadata);
+        if (buffer == null) {
+            // partial write on the header payload
+            // Rewind the channel position to the begining of the file
+            ch.position(0);
+            return null;
+        }
 
-        while (o.hasRemaining()) {
+        if (getChecksum(buffer.array()) != metadata.getPayloadChecksum()) {
+            throw new DataCorruptionException();
+        }
 
-            short magic = o.getShort();
-            channelOffset += Short.BYTES;
+        LogHeader header;
 
-            if (magic != RECORD_DELIMITER) {
-                log.error("Expected a delimiter but found something else while "
-                        + "trying to read file {}", sh.fileName);
-                throw new DataCorruptionException();
-            }
+        try {
+            header = LogHeader.parseFrom(buffer.array());
+        } catch (InvalidProtocolBufferException e) {
+            throw new DataCorruptionException();
+        }
 
-            byte[] metadataBuf = new byte[METADATA_SIZE];
-            o.get(metadataBuf);
-            channelOffset += METADATA_SIZE;
+        return header;
+    }
 
-            try {
-                Metadata metadata = Metadata.parseFrom(metadataBuf);
+    /**
+     * Parse an entry.
+     *
+     * @param ch
+     * @param metadata
+     * @return
+     * @throws IOException
+     */
+    private LogEntry parseEntry(FileChannel ch, Metadata metadata) throws IOException {
+        if (metadata == null) {
+            // The metadata for this entry was partial written
+            return null;
+        }
 
-                byte[] logEntryBuf = new byte[metadata.getLength()];
+        ByteBuffer buffer = getPayloadForMetadata(ch, metadata);
+        if (buffer == null) {
+            // partial write on the entry
+            // rewind the channel position to point before
+            // the metadata field for this partially written payload
+            ch.position(ch.position() - METADATA_SIZE);
+            return null;
+        }
 
-                o.get(logEntryBuf);
+        if (!noVerify) {
+            if (metadata.getPayloadChecksum() != getChecksum(buffer.array())) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)